[JSC] Int52Rep should be aware of NodeBytecodeNeedsNegZero

Constellation opened the pull request:
#### b70d98cee1b977aa3f3e1e57ce6fe57ce4c1b75f
<pre>
[JSC] Int52Rep should be aware of NodeBytecodeNeedsNegZero
<a href="https://bugs.webkit.org/show_bug.cgi?id=289865">https://bugs.webkit.org/show_bug.cgi?id=289865</a>
<a href="https://rdar.apple.com/147147115">rdar://147147115</a>

Reviewed by Yijia Huang.

We can use NodeBytecodeNeedsNegZero information when converting double
to Int52 when Int52Rep is marked as NodeBytecodeNeedsNegZero.

* Source/JavaScriptCore/assembler/testmasm.cpp:
(JSC::testBranchConvertDoubleToInt52):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::arithNodeFlags):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::convertAnyInt):
(JSC::DFG::SpeculativeJIT::speculateAnyInt):
(JSC::DFG::SpeculativeJIT::speculateDoubleRepAnyInt):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileInt52Rep):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::branchConvertDoubleToInt52):

Canonical link: <a href="https://commits.webkit.org/292253@main">https://commits.webkit.org/292253@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/174730a3b9d132195fc32953ac566243e7305e62

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95369 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14972 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4830 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100416 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45874 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15259 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23403 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72735 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30001 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98371 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86086 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53066 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11122 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3820 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45209 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88040 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81298 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3915 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102453 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/93992 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22419 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16361 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81748 "Found 1 new test failure: compositing/patterns/direct-pattern-compositing-size.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22667 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82099 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81122 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25700 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3117 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15700 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15326 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22388 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27527 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/116680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22047 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/116680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23787 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -1518,26 +1518,42 @@ class AssemblyHelpers : public MacroAssembler {
         move(TrustedImm64(0xfff8000000000000ULL), scratch1GPR);
         add64(destGPR, scratch1GPR);
         urshift64(TrustedImm32(52), scratch1GPR);
         failureCases.append(branch64(BelowOrEqual, scratch1GPR, TrustedImm32(4094)));
```
Why do we check `4094` instead of `4095` (`2^12 - 1`)? 

For example, if we check whether `int8` is in `int4` range (from `-8` to `7`), then we do similar things like above
```
7 - 8 = -1 // 1111 1111
-8 - 8 = -16 // 1111 0000
```
then unsigned right shift by 4 bits and both should have
```
0000 1111
```
which means if the result is `<= 15` (`2^4 - 1`) then it's under `int4` range. Similar to check whether `int64` is under `int52` range. We should check `<= 4095` (`2^12 - 1`), right?

Constellation replied:
The range is not -7 to 8. If the number is signed, negative one always has larger range, so the range is -8 to 7.

hyjorc1 replied:
Oops, corrected my previous comment `from -7 to 8` to `from -8 to 7`. But the calculation is correct right?

Constellation replied:
I think `4094` here is right code.
The range in Int52 in hex representation is,

```
negative: 0xfff8000000000000 - 0xffffffffffffffff
positive: 0x0000000000000000 - 0x0007ffffffffffff
```
By adding `0xfff8000000000000`, how they get changed is
```
0xfff8000000000000 => 0xfff0000000000000
0xffffffffffffffff => 0xfff7ffffffffffff
0x0000000000000000 => 0xfff8000000000000
0x0007ffffffffffff => 0xffffffffffffffff
```

This means that the effective range is now shifted to
```
0xfff0000000000000 - 0xffffffffffffffff
```
And when extracting the above bits, then it should be `0xfff0000000000000 >> 52` = `0xfff`
So, if the bit is `0xfff`, then it is right. If it is less than 0xfff, then that's wrong. So `<= 4094` is correct.

hyjorc1 replied:
I see `<= 4094` is for the false case, and `> 4094` is the valid case which `4095` is included. Thanks.

hyjorc1 reviewed (approved):
r=me for this patch, but I have a question for the `4094` boundary check.

Constellation committed:
[JSC] Int52Rep should be aware of NodeBytecodeNeedsNegZero
https://bugs.webkit.org/show_bug.cgi?id=289865
rdar://147147115

Reviewed by Yijia Huang.

We can use NodeBytecodeNeedsNegZero information when converting double
to Int52 when Int52Rep is marked as NodeBytecodeNeedsNegZero.

* Source/JavaScriptCore/assembler/testmasm.cpp:
(JSC::testBranchConvertDoubleToInt52):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::arithNodeFlags):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::convertAnyInt):
(JSC::DFG::SpeculativeJIT::speculateAnyInt):
(JSC::DFG::SpeculativeJIT::speculateDoubleRepAnyInt):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileInt52Rep):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::branchConvertDoubleToInt52):

Canonical link: https://commits.webkit.org/292253@main
